sagemathinc/cocalc
# configurable account registration tokens

# Description
this is for private instances, where admins want to have more control about who can create accounts.

* token: randomized name, optional upper usage limit, optional expiration date, and could be disabled
* ignore the token if it is disabled
* each token has a counter (which is updated upon each successful usage)
   * if the counter is >= the limit, the token is used up
* if the current date is past the expiration, the token can't be used either
* a successful usage of a token is recorded in the central log (+ adding it to the list of PII events)
* show warning if there are no active tokens
* migrate an existing token to the registration table

# Testing Steps
1. before even checking out this branch, create a token (in the site settings)
1. after running the hub, the site settings token should be migrated to the registration token table
1. add one or more token
1. edit a token description, or clear the expiration date or upper limit
1. enable/disable token → check if it blocks
1. expiration date → check if it blocks
1. use up a token to 100% (e.g. 1 with a limit of 1) → check if no longer available
1. delete one or more tokens → no traces are left in the DB
1. check the entry in the central log
1. sorting by expiration date treats the undefined ones as infinitely far in the future

# Relevant Issues

# Deployment

hub first (db update + migration), then webapp

# Screenshot

![Screenshot from 2020-07-03 16-30-27](https://user-images.githubusercontent.com/207405/86478854-538a7080-bd4b-11ea-9e47-d13d6002b09d.png)


### [Checklist](https://github.com/sagemathinc/cocalc/wiki/PR-Checklist):

Front end:
